Providers/Profile.swift

```diff
@@ -911,64 +937,89 @@ public class BrowserProfile: Profile {
         }
 
         /**
-         * Returns nil if there's no account.
+         * Runs the single provided synchronization function and returns its status.
          */
-        private func withSyncInputs<T>(label: EngineIdentifier? = nil, function: (SyncDelegate, Prefs, Ready) -> Deferred<Maybe<T>>) -> Deferred<Maybe<T>>? {
-            if let account = profile.account {
-                if !beginSyncing() {
-                    log.info("Not syncing \(label); already syncing something.")
-                    return deferMaybe(AlreadySyncingError())
-                }
-
-                if let label = label {
-                    log.info("Syncing \(label).")
-                }
-
-                let authState = account.syncAuthState
-
-                let readyDeferred = SyncStateMachine(prefs: self.prefsForSync).toReady(authState)
-                let delegate = profile.getSyncDelegate()
-
-                let go = readyDeferred >>== self.takeActionsOnEngineStateChanges >>== { ready in
-                    function(delegate, self.prefsForSync, ready)
-                }
-
-                // Always unlock when we're done.
-                go.upon({ res in self.endSyncing() })
-
-                return go
+        private func sync(label: EngineIdentifier, function: SyncFunction) -> SyncResult {
+            return syncSeveral([(label, function)]) >>== { statuses in
+                let status = statuses.find { label == $0.0 }
+                return deferMaybe(status!.1)
```

Comment:
I'd feel more comfortable about:

```
return deferMaybe(status?.1 ?? .NotStarted(.Offline))
```

or something. It's possible that `syncSeveral` doesn't return a status for the given label due to a bug, and I'd rather not crash here!